### ISSUE TITLE
mul_div_mod segfaults if compiled without -O3

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#1805]
Using XCode 8.2.1 at commit ff659ac6dc1c04669b016b0e5dd8b8c4c1578007:
```
$ make correctness_mul_div_mod  # defaults to -O3; -O2 and -O1 also succeed
Testing mul vector_width: 1
Testing mul vector_width: 2
Testing mul vector_width: 4
Testing mul vector_width: 8
Testing mul vector_width: 16
Testing div_mod vector_width: 1
Testing div_mod vector_width: 2
Testing div_mod vector_width: 4
Testing div_mod vector_width: 8
Testing div_mod vector_width: 16
Success!

$ OPTIMIZE=-O0 make -j8 correctness_mul_div_mod  # also fails for -g or no flag
Testing mul vector_width: 1
Testing mul vector_width: 2
Testing mul vector_width: 4
Testing mul vector_width: 8
Testing mul vector_width: 16
Testing div_mod vector_width: 1
Testing div_mod vector_width: 2
Testing div_mod vector_width: 4
Testing div_mod vector_width: 8
Testing div_mod vector_width: 16
bash: line 1: 78160 Bus error: 10           /Users/srj/GitHub/Halide/bin/correctness_mul_div_mod
```

**Comment by steven-johnson** [comment]
Quick trip to lldb shows it's stack overflow -- mul_div_mod isn't an unreasonably complex test, though, and I don't think we want to require -O3 to compile all reasonable Halide programs.

**Comment by steven-johnson** [comment]
Failure seems to have been injected by the recent change to mul_div_mod to use std::async(), rather than changes to Simplify itself. Investigating stack-space implications of std::async.

**Comment by steven-johnson** [comment]
Fun facts: 
* in OSX, default stack size for main thread is 8MB; default stack size for secondary threads is 512KB.
* std::thread doesn't offer any way to create a thread with a non-default stack size 
* Our Simplify code can recurse pretty deeply
* Our Simplify code uses a lot less stack space when even simple compiler optimizations are applied

These add up to: some of our testcases happen to use more than 512k of stack in some build configs; it doesn't appear to be a multithreading bug per se (ie there doesn't appear to be a race condition, etc) but rather a side effect of dramatically lower stack size on secondary threads.

**Comment by steven-johnson** [comment]
The TL;DR here is that the vast bulk of the stack space discrepancy is temporary Expr values used as arguments to the calls in IROperator.h/IR.h/Expr.h -- even -O1 will optimize most of them away, but -O0 keeps them all, which more than doubles the stack size of various interesting things in Simplify (e.g. visit(Sub*) goes from ~4k to ~10k stack usage per call for 64-bit code); combine that with stack depth calls in the hundreds and it's not hard to overflow a 512k stack limit. 

While we have generally preferred passing Expr by value (rather than by const-ref), it might be profitable to consider converting many of the calls above to take arguments by const-ref where feasible (which is the vast majority of cases).

**Comment by abadams** [comment]
Probably 17 kinds of undefined behavior, but fun to implement:

https://gist.github.com/abadams/89d7b64139d8faeb9cc0fa6d3a93f982

Changing IROperator.h to use const refs sgtm

**Comment by kgnk** [comment]
You would want a try/catch/rethrow around it. Also, won't compile on MSVC because it doesn't like asm.

**Comment by steven-johnson** [comment]
FWIW: this is now fixed (for compilers that I'm aware of), unless you compile with -fexceptions. (-fexceptions + -O0 still causes stack overflow).

**Comment by abadams** [comment]
A lot of Simplify rules end with expr = mutate(something). These are
effectively tail-recursive calls, but the compiler probably doesn't
interpret them as such. I should rejigger Simplify (at least) to make it so
that you return the new expression instead. It has always been weird that
IR mutators assign to a member instead of returning. All of these would
change to "return mutate(something)", which would presumably be
tail-recursive and not use so much damn stack.

**Comment by steven-johnson** [comment]
Converting a few more heavily-used-in-Simplify calls to const-ref allows us to work even in the with-exceptions case I just identified (see pending PR), but this is probably a fragile peace. Rejiggering to be more stack-friendly would likely be more resilient to future changes.

**Comment by alexreinking** [comment]
Does this still happen?

**Comment by abadams** [comment]
The simplifier has since been changed to use much less stack.
